src/frontend/python/chiquito/dsl.py

```diff
@@ -1,31 +1,50 @@
 from __future__ import annotations
 from enum import Enum
 from typing import Callable, Any
-# import rust_chiquito  # rust bindings
-from chiquito import rust_chiquito
+from chiquito import rust_chiquito  # rust bindings
 import json
-from chiquito import (chiquito_ast, wit_gen)
 
 from chiquito.chiquito_ast import ASTCircuit, ASTStepType, ExposeOffset
 from chiquito.query import Internal, Forward, Queriable, Shared, Fixed
 from chiquito.wit_gen import FixedGenContext, StepInstance, TraceWitness
-from chiquito.cb import Constraint, Typing, ToConstraint, to_constraint
+from chiquito.cb import (
+    Constraint,
+    Typing,
+    ToConstraint,
+    to_constraint,
+    LookupTableRegistry,
+    LookupTable,
+    LookupTableBuilder,
+    InPlaceLookupBuilder,
+)
 from chiquito.util import CustomEncoder, F
 
 
 class CircuitMode(Enum):
     NoMode = 0
     SETUP = 1
     Trace = 2
+    FixedGen = 3
 
 
 class Circuit:
     def __init__(self: Circuit):
         self.ast = ASTCircuit()
+        self.tables = LookupTableRegistry()
         self.witness = TraceWitness()
         self.rust_ast_id = 0
         self.mode = CircuitMode.SETUP
         self.setup()
+        if hasattr(self, "fixed_gen") and callable(self.fixed_gen):
+            self.mode = CircuitMode.FixedGen
+            if self.ast.num_steps == 0:
+                raise ValueError(
+                    "Must set num_steps by calling pragma_num_steps() in setup before calling fixed_gen()."
+                )
+            self.fixed_gen_context = FixedGenContext.new(self.ast.num_steps)
```

Comment:
@qwang98 I don't think we need the class FixedGenContext anymore, we could keep the assignations array in this class

Reply:
Agreed. I kept fixed_assignments in AST as we did in Rust, and added AST function that modifies it. DSL calls this AST function.